File: gridTraveller.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class gridTraveller {

    private Map<String, Integer> memo = new HashMap<>();

    public int traveller(int m, int n){

        String ms = String.valueOf(m);
        String ns = String.valueOf(n);
        String key = ms + ","+ ns;

        if(memo.containsKey(key)){
            return memo.get(key);
        }

        // Base Cases
        if(m == 1 && n == 1){
            return 1;
        }

        if(m == 0 || n == 0){
            return 0;
        }

        int result =  traveller(m-1, n) + traveller(m, n-1);
        memo.put(key, result);

        return memo.get(key);
    }   

    public static void main(String[] args){
        Scanner input = new Scanner(System.in);
        System.out.print("Enter the no. of columns: ");
        int m = input.nextInt();
        System.out.print("Enter the no. of rows: ");
        int n = input.nextInt();
        input.close();

        gridTraveller traveller = new gridTraveller();
        System.out.println(traveller.traveller(m, n));

    }
}

File: gridTraveller.py
def grid_Traveller(m: int, n: int, memo = {}) -> int:
    """
    Returns the number of ways to travel from 
    the start to end point of a grid(mxn)
    """
    # Memoisation
    key = (m, n)

    # Check if key is in memo
    if key in memo:
        return memo[key]

    # Base Case
    if n == 1 and m == 1:
        return 1
    
    # Invalid Grid
    if n == 0 or m == 0:
        return 0

    memo[key] = grid_Traveller(m-1, n, memo) + grid_Traveller(m, n-1, memo)

    return memo[key]

m = int(input('Enter the no. of columns: '))
n = int(input('Enter the no. of rows: '))

print(grid_Traveller(m, n))
